File: plugins/plugin-slack/rust/src/providers/channel_state.rs
//! Channel state provider for Slack.

use crate::service::SlackService;
use crate::types::SlackChannelType;
use serde::{Deserialize, Serialize};

/// Provider name
pub const CHANNEL_STATE_PROVIDER: &str = "slackChannelState";

/// Provider description
pub const CHANNEL_STATE_DESCRIPTION: &str = "Provides information about the current Slack channel context";

/// Channel state data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChannelStateData {
    pub channel_type: String,
    pub workspace_name: String,
    pub channel_name: String,
    pub channel_id: String,
    pub thread_ts: Option<String>,
    pub is_thread: bool,
    pub topic: Option<String>,
    pub purpose: Option<String>,
    pub is_private: Option<bool>,
    pub num_members: Option<i32>,
}

/// Channel state result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChannelStateResult {
    pub data: ChannelStateData,
    pub values: ChannelStateValues,
    pub text: String,
}

/// Channel state values for template substitution
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChannelStateValues {
    pub channel_type: String,
    pub workspace_name: String,
    pub channel_name: String,
    pub channel_id: String,
    pub is_thread: bool,
}

/// Get channel state
pub async fn get_channel_state(
    service: &SlackService,
    channel_id: &str,
    thread_ts: Option<&str>,
    agent_name: &str,
    sender_name: &str,
) -> Result<ChannelStateResult, String> {
    let channel = service
        .get_channel(channel_id)
        .await
        .map_err(|e| e.to_string())?;

    let channel_type = channel.channel_type();
    let channel_type_str = match channel_type {
        SlackChannelType::Im => "DM",
        SlackChannelType::Mpim => "GROUP_DM",
        SlackChannelType::Group => "PRIVATE_CHANNEL",
        SlackChannelType::Channel => "PUBLIC_CHANNEL",
    };

    let channel_name = channel.name.clone();
    let is_thread = thread_ts.is_some();

    let mut response_text = match channel_type {
        SlackChannelType::Im => {
            format!(
                "{} is currently in a direct message conversation with {} on Slack. {} should engage in conversation, responding to messages that are addressed to them.",
                agent_name, sender_name, agent_name
            )
        }
        SlackChannelType::Mpim => {
            format!(
                "{} is currently in a group direct message on Slack. {} should be aware that multiple people can see this conversation.",
                agent_name, agent_name
            )
        }
        _ => {
            if is_thread {
                format!(
                    "{} is currently in a thread within the channel #{} on Slack.\n{} should keep responses focused on the thread topic and be mindful of thread etiquette.",
                    agent_name, channel_name, agent_name
                )
            } else {
                format!(
                    "{} is currently having a conversation in the Slack channel #{}.\n{} is in a channel with other users and should only participate when directly addressed or when the conversation is relevant to them.",
                    agent_name, channel_name, agent_name
                )
            }
        }
    };

    if let Some(ref topic) = channel.topic {
        if !topic.value.is_empty() {
            response_text.push_str(&format!("\nChannel topic: {}", topic.value));
        }
    }

    if let Some(ref purpose) = channel.purpose {
        if !purpose.value.is_empty() {
            response_text.push_str(&format!("\nChannel purpose: {}", purpose.value));
        }
    }

    if let Some(ts) = thread_ts {
        response_text.push_str(&format!("\nThis is a threaded conversation (thread timestamp: {}).", ts));
    }

    Ok(ChannelStateResult {
        data: ChannelStateData {
            channel_type: channel_type_str.to_string(),
            workspace_name: String::new(), // Would need to be passed in
            channel_name: channel_name.clone(),
            channel_id: channel_id.to_string(),
            thread_ts: thread_ts.map(String::from),
            is_thread,
            topic: channel.topic.map(|t| t.value),
            purpose: channel.purpose.map(|p| p.value),
            is_private: Some(channel.is_private),
            num_members: channel.num_members,
        },
        values: ChannelStateValues {
            channel_type: channel_type_str.to_string(),
            workspace_name: String::new(),
            channel_name,
            channel_id: channel_id.to_string(),
            is_thread,
        },
        text: response_text,
    })
}

File: plugins/plugin-slack/typescript/src/providers/channelState.ts
import type { IAgentRuntime, Memory, Provider, State } from "@elizaos/core";
import type { SlackService } from "../service";
import { getSlackChannelType, ServiceType } from "../types";

/**
 * Provider for retrieving Slack channel state information.
 */
export const channelStateProvider: Provider = {
  name: "slackChannelState",
  description: "Provides information about the current Slack channel context",
  get: async (runtime: IAgentRuntime, message: Memory, state: State) => {
    const room = state.data?.room ?? (await runtime.getRoom(message.roomId));
    if (!room) {
      return {
        data: {},
        values: {},
        text: "",
      };
    }

    // If message source is not slack, return empty
    if (message.content.source !== "slack") {
      return {
        data: {},
        values: {},
        text: "",
      };
    }

    const agentName = state?.agentName || "The agent";
    const senderName = state?.senderName || "someone";

    let responseText = "";
    let channelType = "";
    let workspaceName = "";
    let channelName = "";
    const channelId = room.channelId ?? "";
    const threadTs = room.metadata?.threadTs as string | undefined;

    const slackService = await runtime.getService(ServiceType.SLACK) as SlackService;
    if (!slackService || !slackService.client) {
      runtime.logger.warn(
        {
          src: "plugin:slack:provider:channelState",
          agentId: runtime.agentId,
          channelId,
        },
        "No Slack client found",
      );
      return {
        data: {
          room,
          channelType: "unknown",
          channelId,
        },
        values: {
          channelType: "unknown",
          channelId,
        },
        text: "",
      };
    }

    // Get channel info
    const channel = channelId ? await slackService.getChannel(channelId) : null;
    if (channel) {
      channelName = channel.name;
      const slackChannelType = getSlackChannelType(channel);

      if (slackChannelType === "im") {
        channelType = "DM";
        responseText = `${agentName} is currently in a direct message conversation with ${senderName} on Slack. ${agentName} should engage in conversation, responding to messages that are addressed to them.`;
      } else if (slackChannelType === "mpim") {
        channelType = "GROUP_DM";
        responseText = `${agentName} is currently in a group direct message on Slack. ${agentName} should be aware that multiple people can see this conversation.`;
      } else {
        channelType =
          slackChannelType === "group" ? "PRIVATE_CHANNEL" : "PUBLIC_CHANNEL";

        if (threadTs) {
          responseText = `${agentName} is currently in a thread within the channel #${channelName} on Slack.`;
          responseText += `\n${agentName} should keep responses focused on the thread topic and be mindful of thread etiquette.`;
        } else {
          responseText = `${agentName} is currently having a conversation in the Slack channel #${channelName}.`;
          responseText += `\n${agentName} is in a channel with other users and should only participate when directly addressed or when the conversation is relevant to them.`;
        }

        if (channel.topic?.value) {
          responseText += `\nChannel topic: ${channel.topic.value}`;
        }
        if (channel.purpose?.value) {
          responseText += `\nChannel purpose: ${channel.purpose.value}`;
        }
      }
    } else {
      channelType = "unknown";
      responseText = `${agentName} is in a Slack conversation but couldn't retrieve channel details.`;
    }

    // Add workspace context if available
    const teamId = slackService.getTeamId();
    if (teamId && room.worldId) {
      const world = await runtime.getWorld(room.worldId);
      if (world) {
        workspaceName = world.name;
        responseText += `\nWorkspace: ${workspaceName}`;
      }
    }

    // Add thread context if applicable
    if (threadTs) {
      responseText += `\nThis is a threaded conversation (thread timestamp: ${threadTs}).`;
    }

    return {
      data: {
        room,
        channelType,
        workspaceName,
        channelName,
        channelId,
        threadTs,
        isThread: Boolean(threadTs),
        topic: channel?.topic?.value,
        purpose: channel?.purpose?.value,
        isPrivate: channel?.isPrivate,
        numMembers: channel?.numMembers,
      },
      values: {
        channelType,
        workspaceName,
        channelName,
        channelId,
        isThread: Boolean(threadTs),
      },
      text: responseText,
    };
  },
};

export default channelStateProvider;
